C++20 bump; Expose event loop and IP/IPrange objects

dr7ana opened the pull request.

dr7ana committed:
Exposed internal IPRange structs in ip.hpp

dr7ana committed:
constructors

- default ipv4 constructor for lokinet application use
- added Address constructors passing ipv4 and ipv6 objects
- added 001 section focused on exposed ip objects
- moved static_asserts from ip.hpp to ip object unit test in 001

dr7ana committed:
extricated event loop

- event loop extricated into discrete class and membered into Network

dr7ana committed:
clang-tidy move references -> forwarding

dr7ana committed:
Address mutators

- methods to take address and get ipv4/ipv6 objects

dr7ana committed:
ip object methods

- added to_string functionality/formatting for ip{v4,v6} and ip{v4,v6}_net objects
- string formattable for all of the above types
- updated 001 unit tests to reflect correct formatting vs native Address types
- ipv6 string tests are check_nofail for architecture variations in CI
- added comparators for ip{v4,v6} and ip{v4,v6}_net types

jagerman commented on include/oxen/quic/ip.hpp:
```diff
@@ -0,0 +1,176 @@
+    inline constexpr ipv4_net ipv4_loopback = ipv4(127, 0, 0, 1) / 8;
+    inline constexpr ipv6 ipv6_loopback(0, 0, 0, 0, 0, 0, 0, 1);
+
+    inline const std::array ipv4_nonpublic = {
```
Will this work as `constexpr`?

dr7ana replied:
done

dr7ana committed:
ipv6 test variation

- We already disable ipv6 testing on some CI architectures. Passing `--no-ipv6` to CI will skip those tests entirely by flipping the `disable_ipv6` boolean
- Rather than making the ipv6 comparisons strictly CHECK_NOFAIL or bypassing them entirely as we do in other test cases, the `disable_ipv6` boolean's truthiness will determine CHECK vs CHECK_NOFAIL

dr7ana committed:
wrapped deleters

- We already have the logic to create a shared_ptr with a deleter that dispatches destruction to the event loop
- Now we have the ability to create a shared_ptr with a custom deleter that invokes a lambda via the event loop
- This is integral to DNS code and other use-cases that wrap C-style library objects in shared_ptr's. For example, the libevent DNS code requires the constrution of an `::evdns_base` object, which must be freed using `evdns_base_free.` This alone could be passed as a deleter, but the function call requires extra parameters. This wrapped_deleter allows us to invoke whatever 3rd party library functions are required in a thread-safe way

dr7ana committed:
Added ipv{4,6} string ctors plus 001 unit checks

dr7ana committed:
bump oxen-logging: cpp20 fixes

dr7ana committed:
more sensical ipv{4,6} implementations

- reworked constructors
- better unit tests
- windows still sucks
- yatta yatta yatta

dr7ana committed:
sid/experimental fix

dr7ana committed:
added ipv6::to_in6 method

tewinget commented on tests/001-handshake.cpp:
```diff
@@ -93,29 +93,104 @@ namespace oxen::quic::test
+        {
+            uint32_t v4_n;                    // network order ipv4 addr
+            auto v4_h = "192.168.1.1"s;       // host order ipv4 string
+            auto v4_full = "192.168.1.1:0"s;  // full ipv4 addr/port string
```
This could be a non-zero port to have symmetry with the ipv6 section after this.

dr7ana replied:
valid, changed :+1:

dr7ana committed:
Stringformattable concept

- Replaced constexpr IsToStringFormattable with C++20 concept constraining fmt::formatter
- GCC compilers version <=10 need to compile with `-fconcepts`
